routing/payment_session_source.go

```diff
@@ -0,0 +1,150 @@
+package routing
+
+import (
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+// A compile time assertion to ensure MissionControl meets the
+// PaymentSessionSource interface.
+var _ PaymentSessionSource = (*SessionSource)(nil)
+
+// SessionSource defines a source for the router to retrieve new payment
+// sessions.
+type SessionSource struct {
+	// Graph is the channel graph that will be used to gather metrics from
+	// and also to carry out path finding queries.
+	Graph *channeldb.ChannelGraph
+
+	// QueryBandwidth is a method that allows querying the lower link layer
+	// to determine the up to date available bandwidth at a prospective link
+	// to be traversed. If the link isn't available, then a value of zero
+	// should be returned. Otherwise, the current up to date knowledge of
+	// the available bandwidth of the link should be returned.
+	QueryBandwidth func(*channeldb.ChannelEdgeInfo) lnwire.MilliSatoshi
+
+	// SelfNode is our own node.
+	SelfNode *channeldb.LightningNode
+
+	// MissionControl is a shared memory of sorts that executions of payment
+	// path finding use in order to remember which vertexes/edges were
+	// pruned from prior attempts. During payment execution, errors sent by
+	// nodes are mapped into a vertex or edge to be pruned. Each run will
+	// then take into account this set of pruned vertexes/edges to reduce
+	// route failure and pass on graph information gained to the next
+	// execution.
+	MissionControl missionControlInterface
```

Comment:
it's a little strange to have a public struct export a field with a private interface, but not a huge deal. it just won't render properly on the godoc site. does this field need to be exported?

Reply:
It is populated in `server.go`, so the field needs to be exported. The interface itself is more like a package-internal way to be able to inject a mock. I can make the interface itself public, but it wouldn't serve a purpose at the moment.

Reply:
i think it'd be fine to export the interface, a la `MissionController`. Because it has private methods, only types defined in the `routing` package will be able to satisfy it, which fits our use case without breaking godoc

Reply:
Ack on the name `MissionController`. It remains difficult sometimes to find a good name. Interface exported.

Reply:
Exported all methods, as we already start exporting one in the follow up pr #3209.

Reply:
Awesome, latest diff lgtm!